### ISSUE TITLE
Poor ACT Policy Performance After Training (Robot Not Moving or Grasping Properly)

### ISSUE THREAD

**Issue by Vishalsub** [huggingface/lerobot#2213]
Hi,
I’m running ACT training on my custom SO-101 robot setup (based on so101_follower + so101_leader teleop).
After training ~100k steps, my robot’s learned policy shows almost no meaningful motion — it doesn’t approach or grasp the object, even though the dataset and training logs look fine.

I suspect this might be related to training stability (batch size, accumulation, or gradient scaling), but I’d like your advice on how to debug this properly.

lerobot-train \
  --dataset.repo_id=${HF_USER}/record-test \
  --policy.type=act \
  --output_dir=ooutputs/train/act_so101_1 \
  --job_name=act_so101_test \
  --policy.device=cuda \
  --wandb.enable=false \
  --policy.repo_id=${HF_USER}/my_act_policy-v1 \
  --steps 100000 \
  --batch_size 1 \
  --policy.chunk_size=50 \
  --policy.n_action_steps=50 \
  --policy.n_encoder_layers=2 \
  --policy.use_vae=false \
  --policy.use_amp=true \
  --log_freq 50 \
  --eval_freq 100

Hardware:

GPU: RTX 3060 (8 GB VRAM)

OS: Ubuntu 22.04 / CUDA 12.4

Dataset: 49 episodes (~39k frames, two 1080p camera views)

Policy params: ~26 M (ResNet18 backbone, Transformer encoder-decoder)

Expected behavior

After ~30k–100k steps, the ACT policy should at least learn smooth approach or grasp-like motion in replay or real execution.

Actual behavior

Training completes without crash, loss decreases slowly.
During replay (via robot_client.py), robot stays static or moves unpredictably.
Increasing steps doesn’t help much — probably due to very small batch (batch = 1).

Attempted fixes

Enabled --policy.gradient_checkpointing=true
Tried --policy.use_amp=true
Modified training script to simulate gradient accumulation (accum_steps = 16)
Verified dataset integrity (39k frames, correct camera pairs, proper calibration)
Still, the learned policy produces weak or unstable control outputs.

Questions
Does ACT require a minimum batch size (e.g. ≥ 16) for stable gradient updates?
Is there a recommended gradient accumulation approach compatible with lerobot-train?
Could I safely downscale camera images (1080p → 480p) to allow larger batches without hurting performance?
Are there best practices for balancing chunk_size and n_action_steps for small datasets?

**Comment by Yyl-Sylmi** [comment]
I collected 60 rounds of directional data on an RTX5090 device (20 rounds with the target fixed and 40 rounds with the target position slightly moved), with a training step count of 200k and a batch_size of 32. The performance of the trained model is still very poor. The success rate of pointing when the target is fixed is 60%, while the success rate of slightly moving the target position is only 20%，Looking forward to the answer.

**Comment by Yyl-Sylmi** [comment]
> 我在RTX5090设备上收集了60发方向数据（20发固定目标，40发目标位置略微移动），训练步数为20万，batch_size为32。训练模型的性能仍然很差。目标固定时指向的成功率为60%，而稍微移动目标位置的成功率仅为20%，期待答案。

In addition, the same collection and execution of the grasping task performed very well, with success rates of 100% and 60% respectively

**Comment by Yyl-Sylmi** [comment]
My chunk_size, n_action_steps, and n_encoder_layers are all default Settings

**Comment by Rahillasne** [comment]
Batch size 1 is your main problem — ACT needs at least 8 for stable gradients. Downscale your cameras to 480×480 (original ACT used this, no quality loss for tabletop tasks) and you should fit batch=16 on a 3060.
Also: chunk_size=50 and n_action_steps=50 is way too high. Try chunk_size=20, n_action_steps=10. And turn use_vae=true back on — without it, ACT averages contradictory demonstrations into mush.
I ran your dataset format through a quality checker I built (pip install orbit-robotics) — for SO-101 setups with 49 episodes it'll tell you if you have dead joints, action divergence, or consistency issues that explain the failure before you burn more GPU hours. Might save you some debugging time.

**Comment by jashshah999** [comment]
A common cause of "robot not moving" is idle/static frames dominating the dataset, biasing the policy toward zero-action predictions. `lerobot-doctor trim` detects and removes these static frames automatically. Also `lerobot-doctor score` identifies low-quality episodes that might be poisoning training.

```bash
lerobot-doctor trim /path/to/dataset --dry-run
lerobot-doctor score /path/to/dataset
```

https://github.com/jashshah999/lerobot-doctor

**Comment by devika-dudo** [comment]
> Batch size 1 is your main problem — ACT needs at least 8 for stable gradients. Downscale your cameras to 480×480 (original ACT used this, no quality loss for tabletop tasks) and you should fit batch=16 on a 3060. Also: chunk_size=50 and n_action_steps=50 is way too high. Try chunk_size=20, n_action_steps=10. And turn use_vae=true back on — without it, ACT averages contradictory demonstrations into mush. I ran your dataset format through a quality checker I built (pip install orbit-robotics) — for SO-101 setups with 49 episodes it'll tell you if you have dead joints, action divergence, or consistency issues that explain the failure before you burn more GPU hours. Might save you some debugging time.

its 480*640
when i went through the configuration_act.py file in lerobot implementation
they have set chunk_size to a default 100, action_steps to a default 100 

soo if fps is 30 a 100 chunk size would mean the policy have to predict next 100/30s = 3.33s at one timstep?? 

is that why you suggested a lower chunk size and action steps??

according to the actual paper....temporal ensembling should be done at each timestep, right? and a weighted average is taken to get the action at a timestep  ---but in lerobot implementation....we have an option to turn them off and set the timsteps after which next prediction should happen. 



If the chunk is too short, the model cannot plan a complete motion. If too long, the model wastes capacity predicting far-future actions that will be overridden by the next chunk anyway  --- this is what i read in a [blog](https://www.roboticscenter.ai/learn/action-chunking-transformers)
